### PR TITLE
Ad feedback button

### PR DESCRIPTION
**Pull request by lps88**
## What does this change?
Ad slots will have a 'Report ad' button so that we can glean useful feedback from users.

## What is the value of this and can you measure success?
The data will help to identify ads which need to be blocked, and could allow us to track ad quality over time.

## Screenshots
image

image

image

## To Do
- Add support for programmatic ads

NB. this feature is behind a switch for now

### CODE REVIEW

File: static/src/stylesheets/module/_popup.scss [hunk 1]
```diff
@@ -293,3 +293,109 @@ $control-offset: 36 + $gs-gutter/2;
...
+    z-index: $zindex-overlay;
+    padding: 0;
+    ul {
+        margin: 0;
+    }
+    li.popup__item::before {
+        content: none;
+    }
+    h3 {
+        color: #333333;
```
Comment (katebee) [inline static/src/stylesheets/module/_popup.scss]: `color: $neutral-1`

**Comment by NataliaLKB** [comment]
This is great!!! I really like this feature 🎉

File: static/src/stylesheets/module/_popup.scss [hunk 2]
```diff
@@ -293,3 +293,109 @@ $control-offset: 36 + $gs-gutter/2;
...
+        pointer-events: none;
+        width: 0;
+        height: 0;
+        border-bottom: 0;
+        margin-top: 2px;
+        margin-bottom: 2px;
+        margin-left: 4px;
+        border-left: $popup-toggle-arrow-size solid transparent;
+        border-right: $popup-toggle-arrow-size solid transparent;
+        border-top: $popup-toggle-arrow-size solid #005689;
```
Comment (NataliaLKB) [inline static/src/stylesheets/module/_popup.scss]: `$guardian-brand` : http://guardian.github.io/pasteup-palette/demo/

File: static/src/stylesheets/module/_popup.scss [hunk 3]
```diff
@@ -293,3 +293,109 @@ $control-offset: 36 + $gs-gutter/2;
...
+            display: none;
+        }
+        // We just want draw the up arrow instead of the down arrow, to show the (currently open) menu can be closed
+        &:after {
+            content: '';
+            display: inline-block;
+            width: 0;
+            height: 0;
+            border-top: 0;
+            margin-top: 2px;
```
Comment (NataliaLKB) [inline static/src/stylesheets/module/_popup.scss]: There are a lot of duplicate styles in this block and the ones above. Maybe you could do this instead:

```
.ad-feedback.popup__toggle {
    float: right;

    &:after {
        pointer-events: none;
        border-bottom: 0;
        border-top: $popup-toggle-arrow-size solid $guardian-brand;
    }

    &:after,
    &.is-active:after {
        content: '';
        display: inline-block;
        width: 0;
        height: 0;
        margin-top: 2px;
        margin-bottom: 2px;
        margin-left: 4px;
        border-left: $popup-toggle-arrow-size solid transparent;
        border-right: $popup-toggle-arrow-size solid transparent;
    }

    &.is-active {
        &:before {
            display: none;
        }
        
        &:after {
            border-top: 0;
            border-bottom: $popup-toggle-arrow-size solid $guardian-brand;
        }
    }
}
```

File: static/src/stylesheets/module/_popup.scss [hunk 4]
```diff
@@ -293,3 +293,109 @@ $control-offset: 36 + $gs-gutter/2;
...
+
+}
+
+.ad-feedback.popup {
+    width: 240px;
+    top: 18px;
+    left: inherit;
+    right: $gs-gutter;
+    background: $neutral-8;
+    border: 1px solid #dcdcdc;
```
Comment (NataliaLKB) [inline static/src/stylesheets/module/_popup.scss]: `$neutral-4`

File: static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert-label.js [hunk 5]
```diff
@@ -1,14 +1,42 @@
...
+    popupTemplate,
+    config
 ) {
     function renderAdvertLabel(adSlotNode) {
         if (shouldRenderLabel(adSlotNode)) {
+            var zIndexOverlay = 1050;
+            var feedbackPopup = '', feedbackThanksMessage = '';
+            if (config.switches.adFeedback) {
+                adSlotNode.style.zIndex = zIndexOverlay;
+                feedbackPopup = template(popupTemplate, {
```
Comment (regiskuckaertz) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert-label.js]: Any chance you can memoize that? I'm thinking:

- on first run, create a node `labelDiv` with all that stuff, just like you do below
- clone that node (with `deep` flag set to true) and set the ad slot specific data (i.e. just the id)
- insert that clone

File: static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js [hunk 6]
```diff
@@ -156,6 +164,42 @@ define([
                 }) : Promise.resolve();
             }
 
+            function addFeedbackDropdownToggle() {
+                return isRendered ? fastdom.write(function () {
```
Comment (regiskuckaertz) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js]: No switch test?

File: static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js [hunk 7]
```diff
@@ -156,6 +164,42 @@ define([
...
+                    if (!bonzo(advert.node).hasClass('js-toggle-ready')){
+                        new Toggles(advert.node).init();
+                    }
+                }) : Promise.resolve();
+            }
+
+            function applyFeedbackOnClickListeners(slotRenderEvent) {
+                var readyClass = 'js-onclick-ready';
+                return isRendered ? fastdom.write(function () {
+                    bonzo(qwery('.popup__item-problem--option')).each(function(el) {
```
Comment (regiskuckaertz) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js]: The `bonzo` wrapping is unnecessary here: `qwery(',,,').forEach()`.

File: static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js [hunk 8]
```diff
@@ -156,6 +164,42 @@ define([
...
+                        var option = bonzo(el);
+                        if (option.hasClass(readyClass)) return;
```
Comment (regiskuckaertz) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js]: You can get rid of that and just call `qwery('.popup__item-problem--option.js-onclick-ready')` above

File: static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js [hunk 9]
```diff
@@ -156,6 +164,42 @@ define([
...
+                        bean.on(el, 'click', function() {
```
Comment (regiskuckaertz) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js]: `el.addEventListener('click', ...)`

File: static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js (hunk 6 above)
Comment (lps88) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js]: The switch test is in `render-advert-label.js`, would it be better here?

Comment (regiskuckaertz) [inline static/src/javascripts-legacy/projects/commercial/modules/dfp/render-advert.js]: I'm not sure I'm reading this right, but isn't `addFeedbackDropdownToggle` going to run no matter if there is actually a toggle or not? (same for `applyFeedbackOnClickListeners` but I suppose in that case the queries will return empty arrays so we're fine there)

File: static/src/stylesheets/module/_adslot.scss [hunk 10]
```diff
@@ -1,7 +1,7 @@
 .ad-slot {
     position: relative;
     z-index: $zindex-ads;
-    overflow: hidden;
+    overflow: initial;
```
Comment (regiskuckaertz) [inline static/src/stylesheets/module/_adslot.scss]: I understand this one, just slightly worried about side effects. Let's keep an eye when it's out!

**Review by regiskuckaertz — APPROVED** [review]
One small change but otherwise LGTM 👍

**Pull request merged**